app/Server.php

```diff
@@ -93,4 +127,181 @@ public function stats()
     {
         return $this->hasMany(ServerStat::class);
     }
+
+    /**
+     * Scope a query to only get servers that have a description like the passed one.
+     *
+     * @param  Builder $query       Query that should be scoped
+     * @param  String  $description Description to search for
+     * @return Builder The scoped query
+     */
+    public function scopeWithDescription(Builder $query, $description)
+    {
+        return $query->where('description', 'like', '%' . $description . '%');
+    }
+
+    /**
+     * A call to the api failed, mark server as offline
+     * @todo Could be used in the future to trigger alarms, notifications, etc.
+     */
+    public function apiCallFailed()
+    {
+        $this->status     = ServerStatus::OFFLINE;
+        $this->timestamps = false;
+        $this->save();
+    }
+
+    /**
+     * Mark all meetings still marked as running on this server as ended
+     * and cleanup live usage data for corresponding room
+     */
+    public function endMeetings()
+    {
+        foreach ($this->meetings()->whereNull('end')->get() as $meeting) {
+            $meeting->end = date('Y-m-d H:i:s');
+            $meeting->save();
+
+            // If no other meeting is running for this room, reset live room usage
+            if ($meeting->room->runningMeeting() == null) {
+                $meeting->room->participant_count       = null;
+                $meeting->room->listener_count          = null;
+                $meeting->room->voice_participant_count = null;
+                $meeting->room->video_count             = null;
+                $meeting->room->save();
+            }
+        }
+    }
+
+    /**
+     * Panic server, set status offline and try to end all meeting using the api
+     */
+    public function panic()
+    {
+        $this->status = ServerStatus::DISABLED;
+        $this->save();
+
+        $query   = $this->meetings()->whereNull('end');
+        $total   = $query->count();
+        $success = 0;
+        foreach ($query->get() as $meeting) {
+            try {
+                $meeting->endMeeting();
+                $success++;
+            } catch (\Exception $exception) {
```

Comment:
But this would mean that if a meeting couldn't be stopped but the others are running, the other meetings wouldn't be finished?

Reply:
The boolean of the $meeting->endMeeting(); isn't checked as it doen't make a difference (either the meeting will be ended or it is already ended). If the api call fails, it is handled like all other failed api calls (see check, create, join, etc.) -> the server is marked as offline.